contrib/babelfishpg_tsql/src/forjson.c

```diff
@@ -0,0 +1,163 @@
+/*-------------------------------------------------------------------------
+ *
+ * forjson.c
+ *   For JSON clause support for Babel
+ *
+ *-------------------------------------------------------------------------
+ */
+#include "postgres.h"
+
+#include "executor/spi.h"
+#include "fmgr.h"
+#include "utils/guc.h"
+#include "lib/stringinfo.h"
+#include "miscadmin.h"
+#include "parser/parser.h"
+#include "utils/builtins.h"
+#include "utils/json.h"
+
+static StringInfo tsql_query_to_json_internal(const char *query, int mode, bool include_null_value,
+								bool without_array_wrapper,bool root_name_present, const char *root_name);
+static void SPI_sql_row_to_json_path(uint64 rownum, StringInfo result, bool include_null_value);
+
+PG_FUNCTION_INFO_V1(tsql_query_to_json_text);
+
+
+Datum 
+tsql_query_to_json_text(PG_FUNCTION_ARGS)
+{
+	char *query = text_to_cstring(PG_GETARG_TEXT_PP(0));
+	int mode = PG_GETARG_INT32(1);
+	bool include_null_value = PG_GETARG_BOOL(2);
+	bool without_array_wrapper = PG_GETARG_BOOL(3);
+	bool root_name_present = PG_GETARG_BOOL(4);
+	char *root_name = PG_ARGISNULL(5) ? NULL :  text_to_cstring(PG_GETARG_TEXT_PP(5));
+
+	StringInfo result = tsql_query_to_json_internal(query, mode, include_null_value,
+											without_array_wrapper, root_name_present, root_name);
+	
+	PG_RETURN_TEXT_P(cstring_to_text_with_len(result->data, result->len));
+}
+
+
+/*
+ * Map an SQL row to an JSON element, taking the row from the active
+ * SPI cursor.
+ */
+static void
+SPI_sql_row_to_json_path(uint64 rownum, StringInfo result, bool include_null_value)
+{
+	int			i;
+	const char  *sep="";
+	bool 		isnull;
+
+	appendStringInfoChar(result,'{');
+	for (i = 1; i <= SPI_tuptable->tupdesc->natts; i++)
+	{
+		char        *colname;
+		Datum		colval;
+		
+		colname = SPI_fname(SPI_tuptable->tupdesc, i);
+		
+		if (!strcmp(colname,"\?column\?")) /* When column name or alias is not provided */
+		{
+			ereport(ERROR,
+							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							 errmsg("Column expressions and data sources without names or aliases cannot be formatted as JSON text using FOR JSON clause. Add alias to the unnamed column or table")));
+		}
+		
+		colval = SPI_getbinval(SPI_tuptable->vals[rownum],
+							   SPI_tuptable->tupdesc,
+							   i,
+							   &isnull);	
+			
+		if (!isnull)
+		{
+			appendStringInfo(result,sep);
+			sep = ",";
+			appendStringInfo(result, "%s",
+							 tsql_json_build_object(CStringGetDatum(colname), 
+							 						colval,
+							 						SPI_gettypeid(SPI_tuptable->tupdesc,i),false)->data);
+		}
+		else if (include_null_value)
+		{
+			appendStringInfo(result,sep);
+			sep = ",";
+			appendStringInfo(result, "%s",
+							 tsql_json_build_object(CStringGetDatum(colname), 
+							 						colval,
+							 						SPI_gettypeid(SPI_tuptable->tupdesc,i),true)->data);
+		}
```

Comment:
It can be simplified like
```
if (isnull && !include_null_value)
    continue;

appendStringInfo(result,sep);
sep = ",";
appendStringInfo(result, "%s", tsql_json_build_object(..., isnull, ...));
```